File: Json.Tests/assets/Schema/Typescript/UserStore/Friflo.Json.Flow.Auth.Rights.ts
import { TaskType } from "./Friflo.Json.Flow.Sync"
import { Change }   from "./Friflo.Json.Flow.Sync"

export type Right_Union =
    | RightAllow
    | RightTask
    | RightMessage
    | RightSubscribeMessage
    | RightDatabase
    | RightPredicate
;

export abstract class Right {
    abstract type:
        | "allow"
        | "task"
        | "message"
        | "subscribeMessage"
        | "database"
        | "predicate"
    ;
    description: string;
}

export class RightAllow extends Right {
    type:        "allow";
    grant:       boolean;
    description: string;
}

export class RightTask extends Right {
    type:        "task";
    types:       TaskType[];
    description: string;
}

export class RightMessage extends Right {
    type:        "message";
    names:       string[];
    description: string;
}

export class RightSubscribeMessage extends Right {
    type:        "subscribeMessage";
    names:       string[];
    description: string;
}

export class RightDatabase extends Right {
    type:        "database";
    containers:  { [key: string]: ContainerAccess };
    description: string;
}

export class ContainerAccess {
    operations:       OperationType[];
    subscribeChanges: Change[];
}

export type OperationType =
    | "create"
    | "update"
    | "delete"
    | "patch"
    | "read"
    | "query"
    | "mutate"
    | "full"
;

export class RightPredicate extends Right {
    type:        "predicate";
    names:       string[];
    description: string;
}

File: Json.Tests/assets/Schema/Kotlin/src/main/kotlin/UserStore/Friflo.Json.Flow.Auth.Rights.kt

abstract class Right(
    open  val description : String? = null,
)

data class RightAllow(
              val grant       : Boolean,
    override  val description : String? = null,
) : Right (description)

data class RightTask(
              val types       : List<TaskType>,
    override  val description : String? = null,
) : Right (description)

data class RightMessage(
              val names       : List<String>,
    override  val description : String? = null,
) : Right (description)

data class RightSubscribeMessage(
              val names       : List<String>,
    override  val description : String? = null,
) : Right (description)

data class RightDatabase(
              val containers  : HashMap<String, ContainerAccess>,
    override  val description : String? = null,
) : Right (description)

data class ContainerAccess(
              val operations       : List<OperationType>? = null,
              val subscribeChanges : List<Change>? = null,
)

enum class OperationType {
    create,
    update,
    delete,
    patch,
    read,
    query,
    mutate,
    full,
}

data class RightPredicate(
              val names       : List<String>,
    override  val description : String? = null,
) : Right (description)

